### PR TITLE
Analytic imf

### PR DESCRIPTION
**Pull request by tbmiller-astro**
Updated implementation of imf integration to be completely analytic, which is much faster.  Functionality should be identical and you can now use custom (broken) power law IMFs.

### CODE REVIEW

**Review by johnnygreco — CHANGES_REQUESTED** [review]
Thanks so much, Tim! 

Most of my requests are to update the formatting to be more consistent with the rest of the code.  Please make sure the unit tests still pass after the updates. Let me know if have any questions / thoughts!

File: src/artpop/stars/imf.py [hunk 1]
```diff
@@ -292,21 +162,97 @@ def build_galaxy(stellar_mass, num_stars_iter=1e5, **kwargs):
...
         self.m_min = m_min
         self.m_max = m_max
+        self.eval_min = eval_min
         self.num_norm = self.integrate(m_min, m_max, None)
         self.mass_norm = self.m_integrate(m_min, m_max, None)
 
-    def func(self, m):
-        return self._func(m, norm_type=None)
+    def weights(self, mass_grid, norm_type=None,
+        norm_mass_min=None, norm_mass_max=None):
```
[the remark this answers is not in the document]
Comment (tbmiller-astro) [reply src/artpop/stars/imf.py:242]: Added these in the latest commit!

File: src/artpop/stars/imf.py [hunk 2]
```diff
@@ -319,7 +265,32 @@ def integrate(self, m_min=None, m_max=None, norm=False, **kwargs):
...
+
+        if m > b1:
+            return (b0**(a0+1.) - self.eval_min**(a0+1.)) / (a0+1)/c0 \
+             + (b1**(a1+1.) - b0**(a1+1.))/(a1+1)/c1 \
+             + (m**(a2+1.) - b1**(a2+1.))/(a2+1)/c2
+        elif m > b0:
+            return (b0**(a0+1.) - self.eval_min**(a0+1.))/(a0+1)/c0 \
+             + (m**(a1+1.) - b0**(a1+1.))/(a1+1) /c1
+        else:
+            return (m**(a0+1.) - self.eval_min**(a0+1.))/(a0+1)/c0
```
[the remark this answers is not in the document]
Comment (tbmiller-astro) [reply src/artpop/stars/imf.py]: Changed to single `return` statement.

File: src/artpop/stars/isochrones.py [hunk 3]
```diff
@@ -302,8 +302,8 @@ def imf_weights(self, imf, m_min_norm=None, m_max_norm=None,
 
         Parameters
         ----------
-        imf : str or callable function that takes mass as its argument
-            Stellar Initial mass function.
+        imf : str or dict
+            Stellar Initial mass function, see imf.py for details
```
[the remark this answers is not in the document]
Comment (tbmiller-astro) [reply src/artpop/stars/isochrones.py]: Added additional info here and in other functions in isochrones.py and populations.py

File: src/artpop/stars/imf.py [hunk 4]
```diff
@@ -292,21 +162,97 @@ def build_galaxy(stellar_mass, num_stars_iter=1e5, **kwargs):
 class IMFIntegrator(object):
     """
     A helper class for numerically integrating the IMF.
+
+    Parameters
+    ----------
+    prop : str or dict
```
[the remark this answers is not in the document]
Comment (tbmiller-astro) [reply src/artpop/stars/imf.py]: Change the name of all these from `prop` to `params`

File: src/artpop/stars/imf.py [hunk 5]
```diff
@@ -292,21 +162,97 @@ def build_galaxy(stellar_mass, num_stars_iter=1e5, **kwargs):
...
+        'scalo' or 'salpeter'. If dict must contain 'a' with 3 elements describing
+        the slopes of the broken power law and 'b' with two elements describing
+        the locations of the breaks.
+    m_min : float, optional
+        Minimum stellar mass.
+    m_max : float, optional
+        Maximum stellar mass.
+    eval_min : float, optional
+        Technical detail describing where to evaluate the indefinite integral,
+        should not need to change this.
```
[the remark this answers is not in the document]
Comment (tbmiller-astro) [reply src/artpop/stars/imf.py]: I agree, changed it to be hard coded

File: src/artpop/stars/imf.py [hunk 6]
```diff
@@ -8,100 +8,34 @@
...
-    elif norm_type == 'sum':
-        norm = 1 / weights.sum()
+#Can set up similar dict for custom option but need to use 3 segment
+#broken power or hack it (see salpeter below)
+kroupa_prop = {'a':[-0.3,-1.3,-2.3],'b':[0.08,0.5]}
+scalo_prop = {'a':[-1.2,-2.7,-2.3],'b':[1,10]}
 
-    weights *= norm
+#Simple 'hack' to allow for a single power law
+salpeter_prop = {'a':[-2.35]*3,'b':[2e2,3e2]}
```
[the remark this answers is not in the document]
Comment (tbmiller-astro) [reply src/artpop/stars/imf.py]: Added a check in the initialization of the class to check for this. So now can pass a float for a single power law

**Review by johnnygreco — APPROVED** [review]
Assume work – thanks Tim!

**Pull request merged**